remove malloc from solo5

djwillia opened the pull request:
This is part of addressing issue #223.
Most dynamic memory usage was in the virito backend, with some use in IRQ handlers.  

In the IRQ case, I have replace the dynamic memory usage in a queuing implementation to use static buffers.

In the virtio case, I have introduced a new function (`alloc_chunk_4K`) which dynamically adds a chunk of 4K aligned memory to the "bottom of the heap".  This function is only intended to be called before passing control to the libOS, which will see its available memory starting at a dynamically determined point based on how much `alloc_chunk_4K`ed memory was done by Solo5 at initialization time.  (I had originally tried to use large static buffers for virtio, but these dramatically reduced performance on QEMU for reasons I don't understand.)

 The libOS can discover the start location and size of the heap via a new Solo5 call: `solo5_get_info`.

ricarkol commented on kernel/kernel.h:
```diff
@@ -78,6 +78,7 @@ void intr_irq_handler(uint64_t irq);
 
 /* mem.c: low-level page alloc routines */
 void mem_init(void);
+void *alloc_chunk_4K(size_t size);
```
Can you add a comment there saying that this function should only be called before `solo5_app_main`? as doing so will most likely mess with the unikernel heap. In fact, it would be nice to add some assert or some mechanism that ensures that.

ricarkol commented on kernel/mem.c:
```diff
@@ -55,23 +55,28 @@ void mem_init(void)
+    /* for 4K alignment */
+    prev = brk = brk + (0x1000 - (brk & 0xfff));
+
+    brk += size;
```
This can overflow. There should not be a `size` greater than 1k or something like that right? you could assert that.

ricarkol replied:
The name suggests that the memory returned will be rounded up to 4k. No caller is expecting that right?

ricarkol commented on kernel/solo5.h:
```diff
@@ -38,6 +38,16 @@
+ * Platform information.  Currently only memory info.
+ */
+struct solo5_info {
+    char *cmdline;
```
Is the plan to stop passing cmdline thorugh `solo5_app_main`?

mato commented:
I'd prefer a slightly more elaborate set of interfaces:

Instead of:

```
void *alloc_chunk_4K(size_t size);
void solo5_get_info(struct solo5_info *info);
```

I propose:

```
void *mem_ialloc_pages(size_t num);
int solo5_mem_claim_heap(void *heap_start, void *heap_end);
```

Rationale:

1. The private `alloc_chunk_4K()` is now a (limited) page allocator in disguise, so let's just call it that. This also simplifies the implementation, and makes it immediately clear to the user that pages are the unit of allocation (which implies page alignment). The `i` in the `ialloc` refers to the limitation that the function is intended to be called only at _initialisation_ time, i.e. before `solo5_mem_claim_heap()` is called by the libOS. Any later calls to this function will panic (assert), and that should be documented.

2. For the public interface, to me `solo5_get_info()` means nothing by itself and does not reflect the side-effects it should enforce (be called only once and disable the private allocator from that point). Hence the use of `claim` in the name, which should also encourage anyone porting a libOS to look the function up in the documentation to understand the semantics.

What do you think?

mato commented:
Typo in the previous comment -- the public interface should be

```
int solo5_mem_claim_heap(void **start, void **end);
```

The function will return success (0) on the initial call only and failure (-1) on all further calls.

djwillia commented:
@mato I agree with the naming of `ialloc` and the invariant that it should only be called at solo5 initialization time.  

I don't think that the initialization time locking of the heap should be exposed to the user.  What about a private call right before `solo5_app_main` is called that ends initialization?  That way the user visible API doesn't have a weird side effect, and user doesn't need to claim anything.

Then we just need a way to get "information" about where the heap starts and ends.  I had called it info to ideally align it with some of the device info calls.  Right now it is only memory, so we could call it mem_info, but I don't see why it shouldn't be more generic info about the platform.  For example, it might have the number of each type of device.  

@ricarkol sorry I had meant to remove `cmdline` from the `struct solo5_info`; I think it should stay in `solo5_app_main` for now for cmdline parsing reasons.

djwillia commented:
After commit 0eafe28...

@ricarkol I think I have addressed your concerns now.

@mato I've addressed part of your concern with `mem_ialloc_pages`, but don't have a user-facing "claim" call.  I do have an assertion and mechanism to lock the memory layout via `mem_lock_heap`.  What do you think?

ricarkol commented:
@djwillia ACK on the new solo5_info struct (without the cmdline field). LGTM

Kensan commented:
Playing the devil's advocate: since **mem_ialloc_pages** is exclusively required by virtio, is there a way to move it from the general kernel to the virtio-specific implementation? Otherwise all implementations "pay the price" even though the functionality of (page-granular) memory allocation is not needed by ukvm and Muen...

djwillia commented:
@Kensan I would prefer what you suggest and I had actually tried to remove all dynamic memory allocation from Solo5.  In the virtio case there are some potentially large buffers (depending on the backend virtio implementation) and when I had tried with static (large) buffers QEMU started behaving strangely (taking ~5 seconds to boot).  I didn't get to the bottom of that, so this simple dynamic allocation was the simplest I could come up with that did not affect the QEMU boot time.  If you know of a way to remove it and not cause QEMU to freak out I'm all for it.

mato commented:
@djwillia _mem_ialloc_pages()_ needs to do the following, and nothing else:
```
void *mem_ialloc_pages(size_t num)
{
    assert(!mem_locked);
    
    uint64_t prev = heap_start;
    heap_start += num << PAGE_SHIFT;
    assert(heap_start < (unit64_t)&prev);

    return (void *)prev;
}
```

It's an internal function, so we don't care about overflow, only that we're not allocating more memory than we have in low-memory situations. All the stack guard and related code can go, it doesn't make sense here -- I will need to move it into ocaml-freestanding for Mirage.

Regarding your proposed `solo5_get_info()` interface, introducing kitchen-sink style calls like that is almost always a bad idea. I need to think about this some more, but I'm leaning towards supplying the start and end of the "heap" as arguments to `solo5_app_main()` instead.

@Kensan Regarding the cost of the implementation, it's basically those five lines of code above, so not  worth splitting out (and having separate `mem.c`'s again).

mato commented on kernel/mem.c:
```diff
@@ -32,6 +40,10 @@ void mem_init(void)
+
+    /* assert heap is page-aligned (4K boundaries) */
+    assert(!(heap_top & 0xfff));
+    
```
Not needed. heap_start _is_ page aligned.

mato commented on kernel/mem.c:
```diff
@@ -24,6 +24,14 @@ static uint64_t heap_start;
+ * solo5_app_main().
+ */
+static int mem_locked = 0;
+void mem_lock_heap(void) { mem_locked = 1; }
```
Style: Not on one line.

mato commented on kernel/virtio/virtio_blk.c:
```diff
@@ -164,8 +165,10 @@ void virtio_config_block(struct pci_config_info *pci)
     virtq_init_rings(pci->base, &blkq, 0);
 
-    blkq.bufs = calloc(blkq.num, sizeof (struct io_buffer));
+    pgs = (blkq.num * sizeof (struct io_buffer) + PAGE_SIZE - 1) / PAGE_SIZE;
```
Use `((size) - 1 >> PAGE_SHIFT) + 1` to go from `(size)` bytes to pages.

mato commented on kernel/virtio/virtio_blk.c:
```diff
@@ -164,8 +165,10 @@ void virtio_config_block(struct pci_config_info *pci)
+    pgs = (blkq.num * sizeof (struct io_buffer) + PAGE_SIZE - 1) / PAGE_SIZE;
+    blkq.bufs = mem_ialloc_pages(pgs);
     assert(blkq.bufs);
+    memset(blkq.bufs, 0, pgs * PAGE_SIZE);
```
Use `pages << PAGE_SHIFT`.

mato commented on kernel/virtio/virtio_net.c:
```diff
@@ -182,10 +183,15 @@ void virtio_config_network(struct pci_config_info *pci)
     virtq_init_rings(pci->base, &xmitq, VIRTQ_XMIT);
 
-    recvq.bufs = calloc(recvq.num, sizeof (struct io_buffer));
+    pgs = (recvq.num * sizeof (struct io_buffer) + PAGE_SIZE - 1) / PAGE_SIZE;
```
As in `virtio_config_block`, use bit shifting in the same calculations in this function.

mato commented on kernel/virtio/virtio_ring.c:
```diff
@@ -87,14 +87,16 @@ int virtq_add_descriptor_chain(struct virtq *vq,
     vq->num = vq->num_avail = inw(pci_base + VIRTIO_PCI_QUEUE_SIZE);
 
-    data = memalign(4096, VIRTQ_SIZE(vq->num));
+    pgs = (VIRTQ_SIZE(vq->num) + PAGE_SIZE - 1) / PAGE_SIZE;
```
Same here...

Kensan commented:
@djwillia I see, do you still have the static buffer code variant somewhere so I might take a look? Maybe alignment of the **struct io_buffer**s had an adverse effect on performance? That said, I certainly do not want to hold up this PR with my "nitpick".

djwillia commented:
thanks @mato.  I updated to use shifts and simplified `mem_ialloc_pages` as you suggested in cc3117c

I think that passing the `heap_start` and `heap_end` through `solo5_app_main` can cause additional complexity for the ocaml-freestanding case because it means that mirage-solo5 (which implements `solo5_app_main`) must change and we must figure out how to get the information to ocaml-freestanding's sbrk/malloc.  Is your issue with (a) the generality of `solo5_get_info` (which could perhaps be solved by naming it `solo5_get_heap_info` or something), or (b) the existence of the call in the Solo5 API (e.g., getting information about the system that is not passed as a boot parameter)? 

@Kensan I'm afraid I don't have that version lying around.  I didn't even need to use it for `struct io_buffers` to see the issue.  Just the existence of the static buffer was enough to cause my QEMU boot slowness.  If you want to see if QEMU freaks out for you put a big static array of a few MB in size in any of the C files and see what happens to boot time.

mato commented:
> Is your issue with (a) the generality of solo5_get_info (which could perhaps be solved by naming it solo5_get_heap_info or something), or (b) the existence of the call in the Solo5 API (e.g., getting information about the system that is not passed as a boot parameter)?

Both.

(a) is a general issue with naming things, not only related to the problem we are discussing. Calling an API `solo5_get_info()` means nothing to me (what "information"?).

You are correct that (b) is an issue with the existence of the call itself. Thinking about it, if we remove `malloc()` from the Solo5 API then _it does not make sense_ for the application/libOS to run without having that information [about where it's heap should go]. Therefore, it's logical to _always_ pass in as arguments to `solo5_app_main()` and a `solo5_get_info()` call is superfluous.

Regarding modifying the ocaml-freestanding/mirage-solo5 dance, a simple and correct solution would be -- move `solo5_app_main(...)` to ocaml-freestanding, have it call `mirage_app_main(cmdline)` which would be in mirage-solo5.

mato commented:
> Regarding modifying the ocaml-freestanding/mirage-solo5 dance, a simple and correct solution would be -- move solo5_app_main(...) to ocaml-freestanding, have it call mirage_app_main(cmdline) which would be in mirage-solo5.

... or ... even better, don't have a main-type function in mirage-solo5 at all and have ocaml-freestanding call `caml_startup()` directly. Then we just need an interface to get the command line from ocaml-freestanding into mirage-solo5. So this might be an argument for having a `solo5_boot_info()` call or similar. I'll think about it some more.

djwillia commented:
Ah, I went down the route of moving things from mirage-solo5 into ocaml-freestanding and ran into the issue that you mention (related to the command line and the mirage-solo5/ocaml-freestanding relationship).  In the end I thought it was simpler to do it this way, with the added benefit of not needing changes to mirage-solo5 for this PR.  I figured we could figure out the correct relationship between ocaml-freestanding and mirage-solo5 at a later date if necessary.

More generally, I don't think I agree that stuffing everything into arguments to `solo5_app_main` is necessarily the right solution.  I like the symmetry of having a set of `_info` calls about parts of the underlying platform, and I almost think of them like the modules in ukvm.  We have `net_info`, `blk_info`, and I what I had in mind was almost like a `core_info`.

I suggest we rename the function to something that conveys some meaning, but leave the structure as is for now so we don't need to modify the ocaml-freestanding and mirage-solo5 relationship in this PR.  Then we can open an issue to clarify that relationship if we're unhappy with it.  What do you think?

mato commented:
Ok, can we compromise for now on a `mem_info`? I think it should accept/return a `uint64_t *heap_start` and a `size_t *mem_size`, or put them in a `struct` if you want. The reason for `mem_size` rather than a `heap_end` is that the maximum for the heap is dependent on how we manage heap vs. stack, and in this scheme we are relinquishing control of that to the libos layer. Does that work for you?

djwillia commented:
Thanks @mato, that sounds good.  I've updated this one and the accompanying https://github.com/mirage/ocaml-freestanding/pull/29

mato commented:
@djwillia Thanks. This looks fine. Will need to coordinate any merging with #201, so it'll be a bit more time before that happens.

mato commented:
@djwillia Not sure what you did here,  but you appear to have merged master into this branch rather than rebased it on top of master. Can you try and fix (use a temporary local branch to verify the result)?

danlythemanly commented:
@mato thanks; I'm not sure what I did either :)  I think this should be better?

mato commented:
This looks better, thanks!

mato commented:
Superseded by #244.
